core/src/test/scala/org/apache/spark/sql/PartitionTableSuite.scala

```diff
@@ -15,22 +15,435 @@
+    )
+  }
+
+  test("part pruning on year function") {
+    tidbStmt.execute("DROP TABLE IF EXISTS `pt3`")
+    tidbStmt.execute("""
+                       |CREATE TABLE `pt3` (
+                       |  `id` int(11) DEFAULT NULL,
+                       |  `name` varchar(50) DEFAULT NULL,
+                       |  `purchased` date DEFAULT NULL,
+                       |  index `idx_id`(`id`)
+                       |) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
+                       |PARTITION BY RANGE (year(purchased)) (
+                       |  PARTITION p0 VALUES LESS THAN (1995),
+                       |  PARTITION p1 VALUES LESS THAN (2000),
+                       |  PARTITION p2 VALUES LESS THAN (2005),
+                       |  PARTITION p3 VALUES LESS THAN (MAXVALUE)
+                       |)
+                     """.stripMargin)
+    refreshConnections()
+
+    assert(
+      extractDAGReq(
+        spark
+        // expected part info only contains one part which is p.
+          .sql("select * from pt3 where purchased = date'1994-10-10'")
+      ).getPrunedParts
+        .get(0)
+        .getName == "p0"
+    )
+
+    assert(
+      extractDAGReq(
+        spark
+        // expected part info only contains one part which is p2.
+          .sql(
+            "select * from pt3 where purchased > date'1996-10-10' and purchased < date'2000-10-10'"
+          )
+      ).getPrunedParts
+        .get(0)
+        .getName == "p1"
+    )
+
+    assert(
+      {
+        val pDef = extractDAGReq(
+          spark
+          // expected part info only contains two parts which are p0 and p1.
+            .sql("select * from pt3 where purchased < date'2000-10-10'")
+        ).getPrunedParts
+        pDef.size() == 2 && pDef.get(0).getName == "p0" && pDef.get(1).getName == "p1"
+      }
+    )
+
+    assert(
+      {
+        val pDef = extractDAGReq(
+          spark
+          // expected part info only contains one part which is p1.
+            .sql(
+              "select * from pt3 where purchased < date'2005-10-10' and purchased > date'2000-10-10'"
+            )
+        ).getPrunedParts
+        pDef.size() == 1 && pDef.get(0).getName == "p2"
+      }
+    )
+
+    assert(
+      {
+        val pDef = extractDAGReq(
+          spark
+          // or with an unrelated column. All parts should be accessed.
+            .sql(
+              "select * from pt3 where id < 4 or purchased < date'1995-10-10'"
+            )
+        ).getPrunedParts
+        pDef.size() == 4
+      }
+    )
+
+    assert(
+      {
+        val pDef = extractDAGReq(
+          // for complicated expression, we do not support for now.
+          // this will be improved later.
+          spark
+            .sql(
+              "select * from pt3 where year(purchased) < 1995"
+            )
+        ).getPrunedParts
+        pDef.size() == 4
+      }
+    )
+  }
+
+  test("adding part pruning test when index is on partitioned column") {
```

Comment:
Do we need to add a test which partition expression is not a column? e.g., (`a` + 1)